sidecar/supervisor/init.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 IBM Corporation
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package supervisor
+
+import (
+	"os"
+	"os/exec"
+	"os/signal"
+	"syscall"
+
+	"github.com/Sirupsen/logrus"
+)
+
+// Init launches new sidecar process with same args and env vars as current process
+// and runs zombie process cleanup for any inherited orphans
+func Init() {
+	cmd := exec.Command("a8sidecar", os.Args[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Env = os.Environ()
+
+	// listen for SIGCHLDs from any of the supervised processes and any orphans spun off from those
+	go reapZombies()
+
+	go launchChild(cmd)
+
+	// Intercept SIGTERM/SIGINT and stop
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGTERM, syscall.SIGINT)
```

Comment:
Forwarding SIGTERM/SIGINT once is not enough - we need to forward any signal (except for SIGCHLD), any time it's received, as long as the child process is alive, and exit only when the child exists.

I would structure the `Init()` method something along this line:
```go
func Init() {
	cmd := ... // build and set up command object
	
	err := cmd.Start()
	if err != nil {
		shutdown(exitCodeFromError(err))
	}
	
	// No we can launch the signal handler
	go handleSignals(cmd.Process)
	
	err = cmd.Wait()
	shutdown(exitCodeFromError(err))
}

func handleSignals(proc os.Process) {
	// Trap all signals
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan)
	
	for sig := range sigChan {
		switch sig {
			case syscall.SIGCHLD:
				reapZombie()
			default:
				forwardSignal(proc, sig)
		}
	}
}
```